### PR TITLE
task2

### PR DESCRIPTION
**Pull request by echekholskiy**

### CODE REVIEW

File: Task2/Utils/CsvManager.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                csv.WriteField(element);
+                csv.NextRecord();
+            }
+
+            writer.Flush();
+        }
+
+        public void DeleteCsvDirectory()
+        {
+            if (Directory.Exists("CSV"))
```
Comment (asushkov-github) [inline Task2/Utils/CsvManager.cs:38]: "CSV"

File: YandexMarket/Pages/CategoryPage.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using Aquality.Selenium.Forms;
+using OpenQA.Selenium;
+
+namespace YandexMarket.Pages
+{
+    public class CategoryPage : Form
+    {
+        public CategoryPage(string nameOfCategory) 
+            :base(By.XPath($"//h1[contains(text(), '{nameOfCategory}')]"), "CategoryPage"){}
```
Comment (asushkov-github) [inline YandexMarket/Pages/CategoryPage.cs:9]: Please don't use such concrete path with `h1`

File: YandexMarket/Pages/LoginPage.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+﻿
+using Aquality.Selenium.Elements.Interfaces;
+using Aquality.Selenium.Forms;
+using OpenQA.Selenium;
+
+namespace YandexMarket.Pages
+{
+    public class LoginPage : Form
+    {
+        public LoginPage() : base(By.XPath("//*[@class='AuthSocialBlock-title']"), "LoginPage"){}
```
Comment (asushkov-github) [inline YandexMarket/Pages/LoginPage.cs:10]: `//*[@class='AuthSocialBlock']`

File: YandexMarket/Pages/LoginPage.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        private readonly ITextBox _loginTextBox =
+            ElementFactory.GetTextBox(By.XPath("//*[@id='passp-field-login']"), "LoginTextBox");
+
+        private readonly ITextBox _passwordTextBox =
+            ElementFactory.GetTextBox(By.XPath("//*[@id='passp-field-passwd']"), "PasswordField");
+
+        private readonly IButton _logInButton =
+            ElementFactory.GetButton(By.XPath("//*[@class='passp-button passp-sign-in-button']//button"),
```
Comment (asushkov-github) [inline YandexMarket/Pages/LoginPage.cs:19]: `//button[@type='submit']`

**Review by asushkov-github — CHANGES_REQUESTED** [review]

File: Task2/Tests/TestFixture.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        protected string Login;
+        protected string Password;
+
+        [SetUp]
+        public void Initial()
+        {
+            _mainPage = new MainPage();
+            _loginPage = new LoginPage();
+            Login = _credentials.GetValue<string>("login");
+            Password = _credentials.GetValue<string>("password");
```
Comment (asushkov-github) [inline Task2/Tests/TestFixture.cs:27]: Try to use IConfiguration for working with configs
https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration?view=dotnet-plat-ext-5.0

File: YandexMarket/Pages/MainPage.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+            });
+
+            return listOfPopularCategoryLinks;
+        }
+
+        public string GetRandomPopularCategory()
+        {
+            var listOfPopularCategoryNames = GetListOfPopularCategoryNames();
+            var indexOfRandomCategory = new Random().Next(0, listOfPopularCategoryNames.Count - 1);
```
Comment (asushkov-github) [inline YandexMarket/Pages/MainPage.cs:101]: Create `RandomizeExtensions` class and use its extension `listOfPopularCategoryNames.GetRandomValue()`

File: YandexMarket/Pages/MainPage.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public List<string> GetListOfCategoryNames()
+        {
+            var listOfCategoryNames = new List<string>();
+
+            _listOfCategories.ToList().ForEach(category =>
+            {
+                listOfCategoryNames.Add(category.GetText());
+            });
```
Comment (asushkov-github) [inline YandexMarket/Pages/MainPage.cs:124]: `return _listOfCategories.Select(c => c..GetText());`

File: YandexMarket/Pages/MainPage.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            return listOfCategoryNames;
+        }
+
+        public List<string> GetListOfCategoryLinks()
+        {
+            var listOfCategoryLinks = new List<string>();
+
+            _listOfCategories.ToList().ForEach(category =>
+            {
+                listOfCategoryLinks.Add(category.GetAttribute("href"));
```
Comment (asushkov-github) [inline YandexMarket/Pages/MainPage.cs:135]: Create ElementExtensions and use method `category.GetHrefAttribute()`

File: YandexMarket/Pages/MainPage.cs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+            });
+            return listOfCategoryLinks;
```
Comment (asushkov-github) [inline YandexMarket/Pages/MainPage.cs:138]: return _listOfCategories.Select(c => c.GetHrefAttribute());

File: YandexMarket/Pages/MainPage.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+using Aquality.Selenium.Core.Utilities;
+using Aquality.Selenium.Elements.Interfaces;
+using Aquality.Selenium.Forms;
+using OpenQA.Selenium;
+
+namespace YandexMarket.Pages
+{
+    public class MainPage : Form
+    {
+        public MainPage() : base(By.XPath("//h3[contains(text(), 'Покупки на Маркете')]"), "MainPage") {}
```
Comment (asushkov-github) [inline YandexMarket/Pages/MainPage.cs:16]: Put {} braces on new string.

File: Task2/Tests/TestFixture.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+using YandexMarketTests.Utils;
+
+namespace YandexMarketTests.Tests
+{
+    internal class TestFixture
+    {
+        private MainPage _mainPage;
+        private LoginPage _loginPage;
+        private CategoryPage _categoryPage;
+        private readonly JsonSettingsFile _credentials = new(@"..\\..\\..\\..\\Resources\\credentials.json");
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/TestFixture.cs]: could not find a method that returns root catolog. So can use:
Path.Combine(Directory.GetCurrentDirectory(), "..\\..\\..\\Resources\\credentials.json")

File: Task2/Tests/TestFixture.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            _mainPage.OpenMainPage();
+        }
+
+        protected bool MainPageIsOpened()
+        {
+            return _mainPage.State.WaitForDisplayed();
+        }
+
+        protected void Authorization(string login, string password)
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/TestFixture.cs]: done

File: Task2/Tests/TestFixture.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        protected List<string> GetPopularCategoryLinks()
+        {
+            return _mainPage.GetListOfPopularCategoryLinks();
+        }
+
+        protected void WriteListOfCategoriesToLog(List<string> categories)
+        {
+            Logger.Instance.Info(string.Join("", categories)); //TODO перенести мб в утилиты
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/TestFixture.cs]: done

File: Task2/Utils/CsvManager.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+            }
+            else return _instance;
+        }
+
+        public void WriteListToCsv(string fileName, List<string> list)
+        {
+            using var writer = new StreamWriter(Directory.GetCurrentDirectory() + "//CSV//" + fileName + ".csv");
+            using var csv = new CsvHelper.CsvWriter(writer, System.Globalization.CultureInfo.InvariantCulture);
+
+            foreach (var element in list)
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Utils/CsvManager.cs]: done

File: Task2/Utils/CsvManager.cs (hunk 1 above)
Comment (echekholskiy) [reply Task2/Utils/CsvManager.cs]: done

File: YandexMarket/Pages/CategoryPage.cs (hunk 2 above)
Comment (echekholskiy) [reply YandexMarket/Pages/CategoryPage.cs]: done

File: YandexMarket/Pages/LoginPage.cs (hunk 3 above)
Comment (echekholskiy) [reply YandexMarket/Pages/LoginPage.cs]: done

File: YandexMarket/Pages/LoginPage.cs (hunk 4 above)
Comment (echekholskiy) [reply YandexMarket/Pages/LoginPage.cs]: done

File: YandexMarket/Pages/MainPage.cs (hunk 10 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs:16]: done

File: Task2/Tests/YandexTests.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+namespace YandexMarketTests.Tests
+{
+    class YandexTests : TestFixture
+    {
+        [Test]
+        public void YandexTest()
+        {
+            OpenMainPage();
+            var mainPageIsOpened = MainPageIsOpened();
+            mainPageIsOpened.Should().BeTrue("MainPage is not opened");
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/YandexTests.cs]: done

File: Task2/Tests/YandexTests.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            
+            Authorization(Login, Password); 
+            var userIsAuthorized = IsAuthorized();
+            userIsAuthorized.Should().BeTrue("User is not authorized");
+            mainPageIsOpened = MainPageIsOpened();
+            mainPageIsOpened.Should().BeTrue("MainPage is not opened");
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/YandexTests.cs]: done

File: Task2/Tests/YandexTests.cs [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+            var popularCategories = GetPopularCategories();
+            WriteListOfCategoriesToLog(popularCategories);
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/YandexTests.cs]: done

File: Task2/Tests/YandexTests.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+            var categoryPageIsOpened = CategoryIsOpened();
+            categoryPageIsOpened.Should().BeTrue("CategoryPage is not opened");
+
+            OpenMainPage();
+            mainPageIsOpened = MainPageIsOpened();
+            mainPageIsOpened.Should().BeTrue("MainPage is not opened");
+
+            ClickCatalog();
+            var listOfAllCategories = GetAllCategoriesFromCatalog();
+            WriteAllCategoriesToCsvFile(listOfAllCategories);
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/YandexTests.cs]: done

File: Task2/Tests/YandexTests.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            
+            var popularCategoryLinks = GetPopularCategoryLinks();
+            var linksOfAllCategories = GetLinksOfAllCategories();
+            linksOfAllCategories.Should().Contain(popularCategoryLinks,
+                "Сategories from the catalog don't contain all popular categories");
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task2/Tests/YandexTests.cs]: done

File: Task2/Tests/TestFixture.cs (hunk 5 above)
Comment (echekholskiy) [reply Task2/Tests/TestFixture.cs]: done

File: YandexMarket/Pages/MainPage.cs (hunk 6 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs]: done

File: YandexMarket/Pages/MainPage.cs (hunk 7 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs]: done, but it returns IEnumerable, so I've added toList() method.

File: YandexMarket/Pages/MainPage.cs (hunk 8 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs]: done

File: YandexMarket/Pages/MainPage.cs (hunk 9 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs]: done

File: YandexMarket/Pages/MainPage.cs (hunk 10 above)
Comment (echekholskiy) [reply YandexMarket/Pages/MainPage.cs]: done

**Pull request closed without merging**